### ISSUE TITLE
Add support for the ICU message format.

### ISSUE THREAD

**Issue by mwistrand** [dojo/i18n#21]
`dojo-i18n` should support the [ICU message format](http://userguide.icu-project.org/formatparse/messages) for bundle labels. This provides a standard method for specifying how a single message should be formatted given a set of input values (more details about the exact format can be found on the page linked above).

Globalize.js exposes this functionality in its [`messages` module](https://github.com/jquery/globalize#message-module). However, this functionality is internal to the `messages.js` module, so there is no way to do something like this:

``` typescript
i18n(bundle).then(messages => {
  const formatted = formatMessage(messages.birdsCounted, {
    count: 10
  });
});
```

Instead, to access this functionality, all messages must be registered with the Globalize.js ecosystem, with any formatting handled by passing [`Globalize.messageFormatter`](https://github.com/jquery/globalize/blob/master/doc/api/message/message-formatter.md) a namespaced path to a message. Since `dojo-i18n` bundles already require a `bundlePath`, we should be able to use that as the namespace, with bundles being registered within `i18n` as they are loaded:

``` typescript
// dojo-i18n/i18n.ts
export function i18n(bundle, context) {
  // load the locale messages, then register them with Globalize.js
  Globalize.loadMessage({
    [locale]: localeMessages
  });
}

export function formatMessage(key: string, options: { [key: string]: any }, locale?: string): string {
  const globalize = locale ? new Globalize(locale) : Globalize;
  const formatter = globalize.messageFormatter(key);
  return formatter(options);
}
```

This could then be used as:

``` typescript
import i18n, { formatMessage } from 'dojo-i18n/i18n';
import bundle from 'nls/purchases';

i18n(greetings, 'fr').then(() => {
  const message = formatMessage(`${bundle.bundlePath}.recentPurchases`, {
    date: '19 Oct 2016',
    itemCount: 42
  }, 'fr');
});
```

Is there anything else that should be taken into account? Can this API be improved?

**Comment by mwistrand** [comment]
As an aside, while `state.labels` as used by the proposed `dojo-widgets/mixins/createI18nMixin` separates the bundle path and key with a colon, it feels somewhat more natural here to use a period.

**Comment by mwistrand** [comment]
In order to be able to load only the `cldr-data` packages needed for message formatting and pluralization, I had to create a [very basic d.ts](https://github.com/mwistrand/DefinitelyTyped/commit/0850b9142db61543b11121d9ffd73b2630113918) for it. And in order to use Globalize as a UMD package, I also had to update the [Globalize.js d.ts](https://github.com/mwistrand/DefinitelyTyped/commit/927a601f1613101325505e5936784e44cedb029c). Once I have more thoroughly tested those, I will open a PR with DefinitelyTyped.
